lang/udf_cache.hh

```diff
@@ -0,0 +1,208 @@
+        static thread_local stats _stats;
+        return _stats;
+    }
+
+private:
+    using cache_key_type = std::pair<seastar::scheduling_group, cql3::functions::function_name>;
+    struct cache_entry_type {
+        mutable udf_instance inst;
+    };
+    struct udf_cache_entry_size {
+        size_t operator()(cache_entry_type& val) {
+            return val.inst.memory.size(val.inst.store);
+        }
+    };
+    using cache_type = utils::loading_cache<cache_key_type, cache_entry_type, 1, utils::loading_cache_reload_enabled::no, udf_cache_entry_size, utils::tuple_hash, std::equal_to<cache_key_type>>;
+
+public:
+    using key_type = cql3::functions::function_name;
+    using value_type = udf_instance;
+
+private:
+    std::unordered_map<cache_key_type, std::queue<seastar::promise<value_type>>, utils::tuple_hash> _in_use_instances;
+    cache_type _cache;
+    size_t _max_size;
+
+public:
+    explicit udf_cache(size_t size)
+        : _cache(size, std::chrono::seconds(10), wasm_logger)
+        , _max_size(size)
+    {}
+
+private:
+    value_type load(wasm::context& ctx) {
+        auto store = wasmtime::Store(ctx.engine_ptr->get());
+        auto linker = wasmtime::Linker(ctx.engine_ptr->get());
+        auto wasi_def = linker.define_wasi();
+        if (!wasi_def) {
+            throw wasm::exception(format("Setting up wasi failed: {}", wasi_def.err().message()));
+        }
+        auto cfg = wasmtime::WasiConfig();
+        auto set_cfg = store.context().set_wasi(std::move(cfg));
+        if (!set_cfg) {
+            throw wasm::exception(format("Setting up wasi failed: {}", set_cfg.err().message()));
+        }
+        auto instance_res = linker.instantiate(store, *ctx.module);
+        if (!instance_res) {
+            throw wasm::exception(format("Creating a wasm runtime instance failed: {}", instance_res.err().message()));
+        }
+        auto instance = instance_res.unwrap();
+        auto function_obj = instance.get(store, ctx.function_name);
+        if (!function_obj) {
+            throw wasm::exception(format("Function {} was not found in given wasm source code", ctx.function_name));
+        }
+        wasmtime::Func* func = std::get_if<wasmtime::Func>(&*function_obj);
+        if (!func) {
+            throw wasm::exception(format("Exported object {} is not a function", ctx.function_name));
+        }
+        auto memory_export = instance.get(store, "memory");
+        if (!memory_export) {
+            throw wasm::exception("memory export not found - please export `memory` in the wasm module");
+        }
+        auto memory = std::get<wasmtime::Memory>(*memory_export);
+
+        return wasm::udf_instance{.store=std::move(store), .instance=std::move(instance), .func=std::move(*func), .memory=std::move(memory)};
+    }
+
+public:
+    seastar::future<value_type> get(const key_type& key, wasm::context& ctx) {
+        auto key_pair = std::make_pair(seastar::current_scheduling_group(), key_type(key));
```

Comment:
Better to use a struct with named members than force readers to be annoyed at .first and .second.